pkg/controller/jobs/rayjob/rayjob_controller.go

```diff
@@ -0,0 +1,282 @@
+	return j.Status.JobDeploymentStatus != rayjobapi.JobDeploymentStatusSuspended
+}
+
+func (j *RayJob) Suspend() {
+	j.Spec.Suspend = true
+}
+
+func (j *RayJob) ResetStatus() bool {
+	if j.Status.StartTime == nil {
+		return false
+	}
+	j.Status.StartTime = nil
+	return true
+}
+
+func (j *RayJob) GetGVK() schema.GroupVersionKind {
+	return gvk
+}
+
+func (j *RayJob) PodSets() []kueue.PodSet {
+	// len = workerGroups + head
+	podSets := make([]kueue.PodSet, len(j.Spec.RayClusterSpec.WorkerGroupSpecs)+1)
+
+	// head
+	podSets[0] = kueue.PodSet{
+		Name:     "head",
+		Template: *j.Spec.RayClusterSpec.HeadGroupSpec.Template.DeepCopy(),
+		Count:    1,
+	}
+
+	// workers
+	for index := range j.Spec.RayClusterSpec.WorkerGroupSpecs {
+		wgs := &j.Spec.RayClusterSpec.WorkerGroupSpecs[index]
+		replicas := int32(1)
+		if wgs.Replicas != nil {
+			replicas = *wgs.Replicas
+		}
+		podSets[index+1] = kueue.PodSet{
+			Name:     strings.ToLower(wgs.GroupName),
+			Template: *wgs.Template.DeepCopy(),
+			Count:    replicas,
+		}
+	}
+	return podSets
+}
+
+func applySelectors(dst, src map[string]string) map[string]string {
+	if len(dst) == 0 {
+		return src
+	}
+	for k, v := range src {
+		dst[k] = v
+	}
+	return dst
+}
+
+func (j *RayJob) RunWithNodeAffinity(nodeSelectors []jobframework.PodSetNodeSelector) {
+	j.Spec.Suspend = false
+	if len(nodeSelectors) == 0 {
+		return
+	}
+
+	// head
+	headPodSpec := &j.Spec.RayClusterSpec.HeadGroupSpec.Template.Spec
+	headPodSpec.NodeSelector = applySelectors(headPodSpec.NodeSelector, nodeSelectors[0].NodeSelector)
+
+	// workers
+	for index := range j.Spec.RayClusterSpec.WorkerGroupSpecs {
+		workerPodSpec := &j.Spec.RayClusterSpec.WorkerGroupSpecs[index].Template.Spec
+		workerPodSpec.NodeSelector = applySelectors(workerPodSpec.NodeSelector, nodeSelectors[index+1].NodeSelector)
+	}
+}
+
+func cloneSelectors(src map[string]string) map[string]string {
+	dst := make(map[string]string, len(src))
+	for k, v := range src {
+		dst[k] = v
+	}
+	return dst
+}
+
+func (j *RayJob) RestoreNodeAffinity(nodeSelectors []jobframework.PodSetNodeSelector) {
+	// head
+	headPodSpec := &j.Spec.RayClusterSpec.HeadGroupSpec.Template.Spec
+	if !equality.Semantic.DeepEqual(headPodSpec.NodeSelector, nodeSelectors[0].NodeSelector) {
+		headPodSpec.NodeSelector = cloneSelectors(nodeSelectors[0].NodeSelector)
+	}
+
+	// workers
+	for index := range j.Spec.RayClusterSpec.WorkerGroupSpecs {
+		workerPodSpec := &j.Spec.RayClusterSpec.WorkerGroupSpecs[index].Template.Spec
+		workerPodSpec.NodeSelector = cloneSelectors(nodeSelectors[index+1].NodeSelector)
+	}
+}
+
+func (j *RayJob) Finished() (metav1.Condition, bool) {
+	condition := metav1.Condition{
+		Type:    kueue.WorkloadFinished,
+		Status:  metav1.ConditionTrue,
+		Reason:  string(j.Status.JobStatus),
+		Message: j.Status.Message,
+	}
+	finished := false
+	switch j.Status.JobDeploymentStatus {
+	case rayjobapi.JobDeploymentStatusComplete, rayjobapi.JobDeploymentStatusFailedJobDeploy:
```

Comment:
I'm not sure how to handle the `FailedToGetOrCreateRayCluster` condition.

If the condition indicates failed to create RayJob, we should treat it as `Finished`.

wdyt?

Reply:
sure, done